test/parser.spec.js

```diff
@@ -0,0 +1,218 @@
+
+    if (type.isDirectory()) {
+      readDir(entryPath, `${prefix}/${entry}`, output)
+    }
+
+    if (type.isFile()) {
+      output.push({
+        path: `${prefix}/${entry}`,
+        content: fs.createReadStream(entryPath)
+      })
+    }
+  })
+
+  output.push({
+    path: prefix
+  })
+
+  return output
+}
+
+describe('parser', () => {
+  const PORT = 6001
+
+  let ctl
+  let handler = () => {}
+
+  before((done) => {
+    http.createServer((req, res) => {
+      if (req.method === 'POST' && req.headers['content-type']) {
+        handler(req)
+          .then(() => {
+            res.writeHead(200)
+          })
+          .catch(() => {
+            res.writeHead(500)
+          })
+          .then(() => {
+            res.end()
+          })
+
+        return
+      }
+
+      res.writeHead(404)
+      res.end()
+    }).listen(PORT, () => {
+      ctl = APIctl(`/ip4/127.0.0.1/tcp/${PORT}`)
+      done()
+    })
+  })
+
+  describe('single file', () => {
+    const filePath = path.resolve(__dirname, 'fixtures/config')
+    const fileContent = fs.readFileSync(filePath, 'utf8')
+
+    before(() => {
+      handler = async (req) => {
+        expect(req.headers['content-type']).to.be.a('string')
+
+        const files = []
+
+        for await (const entry of parser(req)) {
+          if (entry.type === 'file') {
+            const file = { fileName: entry.name, content: '' }
+
+            for await (const data of entry.content) {
+              file.content += data.toString()
+            }
+
+            files.push(file)
+          }
+        }
+
+        expect(files.length).to.equal(1)
+        expect(files[0].fileName).to.contain('config')
+        expect(files[0].content).to.equal(fileContent)
+      }
+    })
+
+    it('parses ctl.config.replace correctly', (done) => {
+      ctl.config.replace(filePath, (err, res) => {
+        expect(err).not.to.exist
+        done()
+      })
+    })
+
+    it('parses regular multipart requests correctly', async () => {
+      const formData = {
+        file: fs.createReadStream(filePath)
+      }
+
+      await request.post({ url: `http://localhost:${PORT}`, formData: formData })
+    })
+  })
+
+  describe('directory', () => {
+    const dirPath = path.resolve(__dirname, 'fixtures')
+
+    let files = []
+
+    before(() => {
+      handler = async (req) => {
+        expect(req.headers['content-type']).to.be.a('string')
+
+        for await (const entry of parser(req)) {
+          if (entry.type === 'file') {
+            const file = { fileName: entry.name, content: '' }
+
+            for await (const data of entry.content) {
+              file.content += data.toString()
+            }
+
+            files.push(file)
+          }
+        }
+      }
+    })
+
+    beforeEach(() => {
+      files = []
+    })
+
+    it('parses ctl.add correctly', (done) => {
+      ctl.add(readDir(dirPath, 'fixtures'), { recursive: true, followSymlinks: false }, (err, res) => {
```

Comment:
Interactions with the HTTP client are all still callback based in the tests. It would be nice if these were converted as well.

Reply:
Done!